go/cli/mcap/cmd/filter.go

```diff
@@ -327,12 +344,58 @@ func filter(
 			if err != nil {
 				return err
 			}
+			mostRecent, ok := mostRecentMessageBeforeRangeStart[message.ChannelID]
 			if message.LogTime < opts.start {
+				if ok {
+					if mostRecent == nil || mostRecent.LogTime <= message.LogTime {
+						mostRecentMessageBeforeRangeStart[message.ChannelID] = message
+						// Copy the data buffer explicitly, to avoid keeping a reference to the greater
+						// `buf` array that underlies `message.Data`.
+						mostRecentMessageBeforeRangeStart[message.ChannelID].Data = append([]byte{}, message.Data...)
```

Comment:
caveat that I am not super familiar with this filter code and it's been a long time since I looked at it...but...are we assuming here that the file is ordered by log time?   I'm not sure if repeated calls to `lexer.Next(buf)` is giving us in-order reading using the index, but I suspect not. If I'm right, then it means this feature might not do what it claims to do if the file is disordered. I'm not sure what it would take to fix that, but if we don't think it's worth it to fully support disordered files, should we at least try to detect and warn/error when it happens, and maybe also document the limitation in the cli help?